### PR TITLE
fix(i18n): localization fixes

### PR DESCRIPTION
**Pull request by BLumia**
**Short description of changes**

Fix some i18n/l10n issues.

**Context: Fixes an issue?**

No.

**Does this change need documentation? What needs to be documented and how?**

No.

**Status of this Pull Request**

Waiting on second review and style changes

**What is missing until this pull request can be merged?**

2nd review.

## Additional information

In the current state, there are some issues that need to be addressed to make the translation process easies. The issues are:

1. We need to avoid splitting the original strings into many `tr()` if they are from the same sentence.
2. We *may* need to avoid using the term "Country" (recommended to use "Region" or something else instead) to avoid potential [politically controversial term](https://en.wikipedia.org/wiki/Taiwan,_China)s.
3. Recommended to use the 2nd parameter of `tr()` to add a comment to the original strings when translator needs to do extra things.

## Checklist

- [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
- [x] I tested my code and it does what I want
- [x] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#source-code-consistency) 
- [x] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
- [x] I've filled all the content above

### CODE REVIEW

File: src/serverdlg.cpp
Comment (ann0see) [inline src/serverdlg.cpp:196]:
```suggestion
        SystemTrayIcon.setToolTip ( tr ( "%1 server", "%1 is the name of the main application" ).arg( APP_NAME ) );
```
You should run clang-format on your pc this should fix the style.

File: src/soundbase.cpp
Comment (ann0see) [inline src/soundbase.cpp:118]: Not 100$ sure, but I think the error message might be long and even contain HTML. That’s why the message is at the back.

**Review by ann0see — COMMENTED** [review]
Thanks. Good catches. 

Could you please add yourself as translator in src/utils.h?

File: src/serverdlg.cpp
Comment (ann0see) [inline src/serverdlg.cpp:110]: What about "Country/Region" here too?

File: src/util.cpp
Comment (BLumia) [inline src/util.cpp:537]: This would fall `clang-format` check but using the result from `clang-format` will make the style not aligned to the format used above in the same file. What should I do or did I missed something?

File: src/serverdlg.cpp
Comment (BLumia) [reply src/serverdlg.cpp:110]: Also sounds okay to me :)

File: src/util.cpp
Comment (ann0see) [reply src/util.cpp:537]: Hmm. Maybe the clang format versions are different.

Have a look at the patch here: https://github.com/jamulussoftware/jamulus/pull/1927/checks?check_run_id=3154583850

File: src/chatdlg.cpp
Comment (ann0see) [inline src/chatdlg.cpp:135]:
```suggestion
                                     tr ( "Do you want to open the link « %1 » in an external browser?" ).arg ( " <b>" + Url.toString() + "</b> " ),
```

or

```suggestion
                                     tr ( "Do you want to open the link '%1' in an external browser?" ).arg ( " <b>" + Url.toString() + "</b> " ),
```

if that works.

File: src/clientsettingsdlg.cpp
Comment (ann0see) [inline src/clientsettingsdlg.cpp:238]: No need for .setNum() as far as I remember.

File: src/clientsettingsdlg.cpp
Comment (ann0see) [inline src/clientsettingsdlg.cpp:232]: The space should be in this line to remain consistent to other files.

File: src/serverdlg.cpp
Comment (ann0see) [inline src/serverdlg.cpp:118]: What about "Combo box for location of this server"?

File: src/serverdlg.cpp
Comment (ann0see) [inline src/serverdlg.cpp:196]: Do we need the explanation here?

Maybe better make it clear that %1 is a placeholder for the application name (which will be filled automatically) —> no need for the translator to translate it.

File: src/serverdlgbase.ui
Comment (ann0see) [inline src/serverdlgbase.ui:63]: What’s this change?

File: src/serverdlgbase.ui
Comment (ann0see) [inline src/serverdlgbase.ui:223]: Great! Thanks.

**Review by ann0see — CHANGES_REQUESTED** [review]

File: src/serverdlgbase.ui
Comment (BLumia) [reply src/serverdlgbase.ui:63]: I‘m not quite sure since it was changed by the QtDesigner. Not sure if it will break something or not if I manually change it back.

File: src/serverdlg.cpp
Comment (BLumia) [reply src/serverdlg.cpp:196]: Translators will only see "%1 server" which can be confused if no context is provided, it can be helpful if we switch to other translation platforms like Transifex, Crowdin, or Weblate later. I guess it's okay for the current translation workflow (translator will need to clone the repo and Linguist can provide contexts like UI or code.

File: src/soundbase.cpp
Comment (BLumia) [reply src/soundbase.cpp:118]: It's no different from the older form in action so I guess it's okay?

Comment (ann0see) [reply src/soundbase.cpp:118]: Argh, yes. So it’s ok probably.

File: src/chatdlg.cpp
Comment (ann0see) [inline src/chatdlg.cpp:135]: The question is if single quotes work…

Comment (BLumia) [reply src/chatdlg.cpp:135]: If you mean if single quotes can be used inside `tr()` with `%1`, yes it will work fine.

``` cpp
// output will be: "aaa 'bbb' ccc"
qDebug() << tr("aaa '%1' ccc").arg("bbb");
```

Comment (ann0see) [reply src/chatdlg.cpp:135]: Ok. Going to trust you on that ;-).

Comment (ann0see) [reply src/chatdlg.cpp:135]:
```suggestion
                                     tr ( "Do you want to open the link '%1' in an external browser?" ).arg ( "<b>" + Url.toString() + "</b>" ),
```

File: src/clientsettingsdlg.cpp
Comment (ann0see) [inline src/clientsettingsdlg.cpp:234]: For this file see: https://github.com/jamulussoftware/jamulus/pull/1889

File: src/clientsettingsdlg.cpp
Comment (ann0see) [inline src/clientsettingsdlg.cpp:270]: Good idea.

File: src/clientsettingsdlg.cpp
Comment (ann0see) [inline src/clientsettingsdlg.cpp:725]: Is that the modern way?

File: src/serverdlg.cpp
Comment (ann0see) [inline src/serverdlg.cpp:111]:
```suggestion
                                 "which this server is located can be set here. If a location is "
```

File: src/util.cpp
Comment (ann0see) [inline src/util.cpp:537]: Hmm. This looks strange. But as far as I know, clang-format complained if you did it as above?

File: src/clientsettingsdlg.cpp
Comment (BLumia) [reply src/clientsettingsdlg.cpp:234]: Could do a rebase once that PR get merged :)

File: src/util.cpp
Comment (BLumia) [reply src/util.cpp:537]: > clang-format complained if you did it as above?

[Yes](https://github.com/jamulussoftware/jamulus/pull/1927#discussion_r676131598)

File: src/clientsettingsdlg.cpp
Comment (BLumia) [reply src/clientsettingsdlg.cpp:725]: I actually never used `QString::setNum()` before.

Comment (ann0see) [reply src/clientsettingsdlg.cpp:725]: MAYBE you don't even need ::number. At least that might be possible.

File: src/serverdlgbase.ui
Comment (ann0see) [inline src/serverdlgbase.ui:222]: Should be documented somewhere else (for the next release): Translators should add this in their language.

File: src/clientsettingsdlg.cpp
Comment (BLumia) [reply src/clientsettingsdlg.cpp:725]: Really? I don't see any documented API for things like `QString(int number)` or similar. You can pass a number to `QString()` because there is a `QString(QChar ch)` but that's not for this use case. For example, `QString(5)` will results in a Unicode character "\u0005", NOT a string with a number "5". `::setNum()` or `QString::number()` should be used instead,

Comment (ann0see) [reply src/clientsettingsdlg.cpp:725]: Ok. So my assumption was wrong. Probably an issue with my PHP background ;-).

**Review by ann0see — APPROVED** [review]
All in all, I think it’s ok. Style wise it’s a bit odd (especially your addition of your name as discussed above). 

Nevertheless: waiting on review from another person.

**Comment by ann0see** [comment]
Todo:

- [ ] QString::number is new (and not yet used anywhere in the code, as far as I know). I‘d accept it. Not sure about others
- [ ] Compilation check on other OS (needed?)
- [ ] Finally: Squash & merge.

**Comment by BLumia** [comment]
> Style wise it’s a bit odd (especially your addition of your name as discussed above).

I can change back to what I did in 87f6770 if preferred as I originally did, but `clang-format` would fail after that.

> QString::number is new (and not yet used anywhere in the code, as far as I know). I‘d accept it. Not sure about others

I can change back to use `QString().setNum()` if preferred as I originally did. I change it to QString::number because you say [it's no need for .setNum()](https://github.com/jamulussoftware/jamulus/pull/1927#discussion_r676139025) so I'm a little confused since it's the only other way to do :/

btw, the *Allow edits by maintainers* option is enabled so feel free to push changes directly to my patchset :)

**Comment by ann0see** [comment]
> but clang-format would fail after that.

That's why we can't change that. So we need to leave it as is.

> I can change back to use QString().setNum()

I think in the comment you mentioned I wasn't clear enough, we usually use:

```
.arg ( APP_NAME )
.arg ( SYSTEM_SAMPLE_RATE_HZ )
```
See: https://github.com/jamulussoftware/jamulus/blob/7fa437292dcfd61d08e4fbb9ebd1cb0c500cd0e4/windows/sound.cpp#L161-L162

However, this does *not* mean the current style cannot be improved.

> btw, the Allow edits by maintainers option is enabled so feel free to push changes directly to my patchset :)

Thanks for the information. Usually we (or at least I) avoid doing that since it might cause confusion.

**Comment by ann0see** [comment]
BTW @BLumia have a look at https://github.com/jamulussoftware/jamulus/releases/tag/r3.8.0devNightly1 this should include your translation.

**Comment by BLumia** [comment]
> > I can change back to use QString().setNum()
> 
> I think in the comment you mentioned I wasn't clear enough, we usually use:
> 
> ```
> .arg ( APP_NAME )
> .arg ( SYSTEM_SAMPLE_RATE_HZ ) // note there's no QString.setNum() here.
> ```
> 
> See:
> 
> https://github.com/jamulussoftware/jamulus/blob/7fa437292dcfd61d08e4fbb9ebd1cb0c500cd0e4/windows/sound.cpp#L161-L162
> 
> However, this does _not_ mean the current style cannot be improved.
 
Oh! I didn't realize you mean chaining `.arg()`. That's also okay and let me know if chaining `arg()` is preferred :)

**Comment by ann0see** [comment]
Hmm. I think chaining .arg() is what we used till now. So I think this should be preferred.

**Review by ann0see — APPROVED** [review]
Thank you!

File: src/clientsettingsdlg.cpp
Comment (ann0see) [inline src/clientsettingsdlg.cpp:727]: Maybe this could be done with arg() too. 
```suggestion
    lblNetBuf->setText ( QString ( tr ( "Size: %1" ) ) .arg ( iCurNumNetBuf ) );
```

Similar below. 

But I‘m not a Qt expert.

**Comment by ann0see** [comment]
Tagging @gilgongo for a review here since it concerns some language changing/…

File: src/clientsettingsdlg.cpp
Comment (BLumia) [reply src/clientsettingsdlg.cpp:727]: Just `tr("").args()` will be fine :)

```suggestion
    lblNetBuf->setText ( tr ( "Size: " ).arg ( iCurNumNetBuf );
```

File: src/chatdlg.cpp
Comment (gilgongo) [inline src/chatdlg.cpp:135]: Is "your" a bit less technical?

```suggestion
                                     tr ( "Do you want to open the link '%1' in your browser?" ).arg ( "<b>" + Url.toString() + "</b>" ),
```

File: src/clientsettingsdlg.cpp
Comment (gilgongo) [inline src/clientsettingsdlg.cpp:233]: Is "works" clearer than merely "supports" (it removes the slighty awkward repetition of "support" too)?

```suggestion
                                           "currently only works with devices supporting a sample rate of %2"
```

File: src/serverdlg.cpp
Comment (gilgongo) [inline src/serverdlg.cpp:111]: As per style guide, we should try to put things into active case for better readability. So this would be:

"Set the country or region in which this server is located here. Clients will show this location in their connect dialog's server list."

File: src/clientsettingsdlg.cpp
Comment (ann0see) [inline src/clientsettingsdlg.cpp:732]: Should be as above (.arg()).

File: src/serverdlg.cpp
Comment (gilgongo) [inline src/serverdlg.cpp:118]: Best practice for copy is not to refer to UI or positional elements as they might change (and in this case people might not know what a "combo box" is). Can we just say this instead? I don't know where this text appears exactly though.

```suggestion
    cbxLocationCountry->setAccessibleName ( tr ( "The location of this server" ) );
```

File: src/serverdlgbase.ui
Comment (gilgongo) [reply src/serverdlgbase.ui:222]: "suffix" is a bit technical perhaps.

```suggestion
            <string extracomment="Consider adding &quot;(Lang)&quot; at the end so that users who select the wrong language can find the correct button">Language</string>
```

**Comment by BLumia** [comment]
About suggest changes: `supports -> work with` and `an external -> your`:

This patch is intended as a fix for the hard-to-translate separated strings instead of an improvement of original copywriting so I kept the original copywriting as-is unchanged (expect the region thing). Do we really need to tweak the other copywriting in this patch? I believe you can find more issues in the original copywriting strings if you do a review for other strings but IMO it's not related to the topic of this patch.

Let me know if you insist to do these changes in this patch set.

File: src/serverdlg.cpp
Comment (BLumia) [reply src/serverdlg.cpp:118]: These strings set with `setAccessibleName()` are for screen readers, they are actually not displayed on the UI.

File: src/clientsettingsdlg.cpp
Comment (BLumia) [reply src/clientsettingsdlg.cpp:732]: Oh, I missed this one, good catch :)

File: src/clientsettingsdlg.cpp
Comment (BLumia) [inline src/clientsettingsdlg.cpp:731]:
```suggestion
    lblNetBufServer->setText ( tr ( "Size: %1" ).arg ( iCurNumNetBufServer ) );
```

File: src/serverdlg.cpp
Comment (BLumia) [inline src/serverdlg.cpp:113]: > As per style guide, we should try to put things into active case for better readability. So this would be:
>
> "Set the country or region in which this server is located here. Clients will show this location in their connect dialog's server list."

@gilgongo is this correct?

```suggestion
                            tr ( "Set the country or region in which this server is located here. "
                                 "Clients will show this location in their connect dialog's server "
                                 "list." );
```

File: src/serverdlg.cpp
Comment (BLumia) [reply src/serverdlg.cpp:111]: See [here](https://github.com/jamulussoftware/jamulus/pull/1927/files#r684075970).

**Comment by gilgongo** [comment]
> This patch is intended as a fix for the hard-to-translate separated strings

Oh OK, I was just respondind to @ann0see tag as that implied these were new strings.

**Comment by ann0see** [comment]
Yes, although there are no new strings, it’s still good to have the old ones reviewed from a UX POV. 
Maybe not part of this PR then.

**Comment by ann0see** [comment]
@BLumia Could you please rebase this PR since mine was merged?

**Comment by BLumia** [comment]
Squashed the patches locally and rebased them onto the master, done :)

File: src/serverdlg.cpp
Comment (pljones) [reply src/serverdlg.cpp:113]: > server is located

"where the server is running" might be more explicit as to what's actually useful.

I've seen people putting the location the server is expected to be used from (e.g. "Cornwall" for a server that I'd say is almost certainly hosted in London, unless they've got a dedicated fibre from the data centre Jamulus Explorer runs in...  Bristol, which is closer and does have good network links, has a higher ping...).

**Comment by ann0see** [comment]
Thanks

File: src/serverdlg.cpp
Comment (BLumia) [reply src/serverdlg.cpp:113]:
```suggestion
                            tr ( "Set the country or region where the server is running. "
                                 "Clients will show this location in their connect dialog's server "
                                 "list." );
```

@pljones is this correct?

Comment (gilgongo) [reply src/serverdlg.cpp:113]: Looks good to me.

**Review by gilgongo — APPROVED** [review]
Looks good from my POV.

File: src/chatdlg.cpp
Comment (pljones) [inline src/chatdlg.cpp:135]: I hadn't realised `tr()` returns a QString that can then have `arg()` applied.  This is a nice idiom that could be used further.

File: src/serverdlg.cpp
Comment (pljones) [reply src/serverdlg.cpp:117]: Most of the accessibility string state the nature of the UI component that they are describing - it helps the visually impared handle the screen navigation if they're prompted in this way.

File: src/serverdlgbase.ui
Comment (pljones) [reply src/serverdlgbase.ui:63]: It usually happens if you save when exiting the UI Designer having examined a field and not left it in its "clean" state.  I think it should be reverted.

**Comment by pljones** [comment]
There's a tiny revert I'd like done but it looks good other than that.

File: src/chatdlg.cpp
Comment (BLumia) [reply src/chatdlg.cpp:135]: This patchset contains all the strings I found while doing the Simplified Chinese translation that concatenates use `+` instead of a formatted string with `arg()` (since it causes trouble for translating). I'm not sure if I missed some other places but most major strings I found while translating are covered in this PR.

File: src/serverdlgbase.ui
Comment (ann0see) [reply src/serverdlgbase.ui:63]: @BLumia could you please revert this as Peter suggested?

Comment (BLumia) [reply src/serverdlgbase.ui:63]: > It usually happens if you save when exiting the UI Designer having examined a field and not left it in its "clean" state.

Yeah as you can see I did add the `extracomment` field with the designer mode. I didn't touch any other thing so it's probably caused by that. anyway here we go :)

```suggestion
      <number>0</number>
```

**Pull request merged**
